<code>Data Cleaning</code>

In [2]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.3.1 -> 23.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import pandas as pd

In [4]:
# read data

data = pd.read_csv('./HousingData.csv') # boston housing data from Kaggle
data.tail()

CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
501  0.06263  0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273     21.0   
502  0.04527  0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273     21.0   
503  0.06076  0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273     21.0   
504  0.10959  0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273     21.0   
505  0.04741  0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1  273     21.0   

          B  LSTAT  MEDV  
501  391.99    NaN  22.4  
502  396.90   9.08  20.6  
503  396.90   5.64  23.9  
504  393.45   6.48  22.0  
505  396.90   7.88  11.9

- CRIM: per capita crime rate by town
- ZN: proportion of residential land zoned for lots over 25,000 sq.ft
- INDUS: proportion of non-retail business acres per town
- CHAS: Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
- NOX: nitric oxides concentration (parts per 10 million)
- RM: average number of rooms per dwelling
- AGE: proportion of owner-occupied units built prior to 1940
- DIS: weighted distances to five Boston employment centers
- RAD: index of accessibility to radial highways
- TAX: full-value property-tax rate per $10,000
- PTRATIO: pupil-teacher ration by town
- B: 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
- LSTAT: % lower status of the population
- MEDV: Median value of owner-occupied homes in $1000's

<code>Missing Values</code>

In [5]:
data.shape

(506, 14)

And that shows - there are 506 rows begins with 0 with 14 columns

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     486 non-null    float64
 1   ZN       486 non-null    float64
 2   INDUS    486 non-null    float64
 3   CHAS     486 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      486 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    486 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


As we can see, the column CRIM, ZN, INDUS, CHAS, AGE, LSTAT have some missing values and the number of missing values are (506 - 486) 20.

<code>Analyzing Outliers</code>

- Boxplots are extremely useful to analyze outliers
- Boxplots displays most of the details that we need such as mean, IQR, Skewness, median in dataset using plotly

In [7]:
pip install seaborn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.3.1 -> 23.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [8]:
pip install -U matplotlib

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.3.1 -> 23.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [9]:
pip install plotly

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.3.1 -> 23.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [10]:
import plotly.express as px

fig = px.box(data, y='RM')
fig.show()

- The blue dots are represents the outliers, the number below and above: 4.8 and 7.7 respectively.
- The room number (RM) represents the bottom and top line of the blue box.


In [11]:
data.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  486.000000  486.000000  486.000000  486.000000  506.000000  506.000000   
mean     3.611874   11.211934   11.083992    0.069959    0.554695    6.284634   
std      8.720192   23.388876    6.835896    0.255340    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.081900    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.253715    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.560263   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  486.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.518519    3.795043    9.549407  408.237154   18.455534  356.674032   
std     27.999513    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.175000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     76.800000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     93.975000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  486.000000  506.000000  
mean    12.715432   22.532806  
std      7.155871    9.197104  
min      1.730000    5.000000  
25%      7.125000   17.025000  
50%     11.430000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

The median, first quartile and third quartile matches using both plotly and describe function.

<code>Handling Missing Values</code>

- Missing values often found in many data analysis and the Python library Pandas enables data analyst or scientist make easier to perform data cleaning and preparation without any pain.
- Pandas method dropna() allows us to remove missing values and fillna() allows us to replace missing value to NA.
- Pandas to_numeric values to missing values which can be imputed with fillna() or removed with dropna().

Before we saw, there are few missing values. Now the time has arrived to handle missing values in the dataset. To handle missing values, we use Pandas dropna() function.

In [12]:
data.dropna()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222   
5    0.02985   0.0   2.18   0.0  0.458  6.430  58.7  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
499  0.17783   0.0   9.69   0.0  0.585  5.569  73.5  2.3999    6  391   
500  0.22438   0.0   9.69   0.0  0.585  6.027  79.7  2.4982    6  391   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
5       18.7  394.12   5.21  28.7  
..       ...     ...    ...   ...  
499     19.2  395.77  15.10  17.5  
500     19.2  396.90  14.33  16.8  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  

[394 rows x 14 columns]

Now, in total there are 394 rows after removing all the missing values in the dataset. We have to fill these missing values using statistics, mean which provides similar numeric values to all the missing data. We have six attributes that have missing values.

In [13]:
data['CRIM'].fillna(data['CRIM'].mean(), inplace=True)
data['ZN'].fillna(data['CRIM'].mean(), inplace=True)
data['INDUS'].fillna(data['CRIM'].mean(), inplace=True)
data['CHAS'].fillna(data['CRIM'].mean(), inplace=True)
data['AGE'].fillna(data['CRIM'].mean(), inplace=True)
data['LSTAT'].fillna(data['CRIM'].mean(), inplace=True)

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


We got all the missing data to some numeric value.

In [15]:
data['AGE']

0      65.200000
1      78.900000
2      61.100000
3      45.800000
4      54.200000
         ...    
501    69.100000
502    76.700000
503    91.000000
504    89.300000
505     3.611874
Name: AGE, Length: 506, dtype: float64

In [16]:
data['AGE'].mean()

65.95303849688513

<code>Handling Bad values</code>

Another interesting aspect of data cleaning is handling bad values. Often it doesn't make any sense so to illustrate this, let's look at the AGE attribute in our original dataset where AGE dtype is float64 and sometimes it is in string data type. So, we have to change it to integers.

In [17]:
df = pd.read_csv('./HousingData.csv')

In [18]:
print(df['AGE'].mean())

68.51851851851852


In [25]:
print(df['AGE'].mean())

68.51851851851852


We have to change Age data type float to integers. For thi, we have to change to integer data type.

In [27]:
df['AGE'] = pd.to_numeric(df['AGE'])
print(df['AGE'].mean())

68.51851851851852


In [29]:
df['AGE'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 506 entries, 0 to 505
Series name: AGE
Non-Null Count  Dtype  
--------------  -----  
486 non-null    float64
dtypes: float64(1)
memory usage: 4.1 KB
